slides/rules-of-rspec.markdown

```diff
@@ -0,0 +1,246 @@
+  
+--
+
+### Here is a simple example that illustrates when each hook is called.
+
+```ruby
+describe "Before and after hooks" do 
+   before(:each) do 
+      puts "Runs before each Example" 
+   end 
+   after(:each) do 
+      puts "Runs after each Example" 
+   end 
+   before(:all) do 
+      puts "Runs before all Examples" 
+   end 
+   after(:all) do 
+      puts "Runs after all Examples"
+   end 
+
+   it 'is the first Example in this spec file' do 
+      puts 'Running the first Example' 
+   end 
+
+   it 'is the second Example in this spec file' do 
+      puts 'Running the second Example' 
+   end 
+end
+```
+
+```
+Runs before all Examples 
+Runs before each Example 
+Running the first Example 
+Runs after each Example 
+.Runs before each Example 
+Running the second Example 
+Runs after each Example 
+.Runs after all Examples
+```
+
+--
+
+The syntax of around is similar to that of `before` and `after` but the semantics are quite different. `before` and `after` hooks are run in the context of of the examples with which they are associated, whereas `around` hooks are actually responsible for running the examples. Consequently, `around` hooks do not have direct access to resources that are made available within the examples and their associated `before` and `after` hooks.
+
+
+```ruby
+RSpec.describe "around hook" do
+  around(:example) do |example|
+    puts "around example before"
+    example.run
+    puts "around example after"
+  end
+
+  it "gets run in order" do
+    puts "in the example"
+  end
+end
+```
+Output should contain:
+
+```
+around example before
+in the example
+around example after
+```
+
+---
+
+## Database cleaning
+
+The name of this setting is a bit misleading. What it really means in Rails
+is "run every test method within a transaction." In the context of rspec-rails,
+it means "run every example within a transaction."
+
+The idea is to start each example with a clean database, create whatever data
+is necessary for that example, and then remove that data by simply rolling back
+the `transaction` at the end of the example.
+
+--
+
+### Setup Database cleaning
+
+If you are using Rails earlier than 5.1. The first thing we need to do is add our gem to Gemfile in `:test` group. Because selenium-webdriver gem, is installed by default in Rails 5.1.
+
+```ruby
+group :test do
+  gem 'selenium-webdriver'
+end
+```
+
+Make sure that `use_transactional_fixtures = true` it will run every example within a transaction
+
+spec/rails_helper.rb <!-- .element: class="filename" -->
+
+```ruby
+config.use_transactional_fixtures = true
+```
+
+Create a `basic_configure`, which manages the Capybara driver configuration:
+
+spec/support/basic_configure.rb <!-- .element: class="filename" -->
+
+```ruby
+RSpec.configure do |config|
+  config.before(:each, type: :system) do
+    driven_by :rack_test
+  end
+
+  config.before(:each, type: :system, js: true) do
+    driven_by :selenium_chrome_headless
+  end
+end
```

Comment:
Are you sure that it is default capybara config for db_clean? And may be describe what is `type: :system, js: true`, what is `driven_by :rack_test` and `selenium_chrome_headless`?